File: Project_2/Sequential/KSelect.py
import numpy as np
import random


def partitionInPlace(A, v):
    n = len(A)
    
    i = 0
    j = n - 1
    
    s = i
    e = j
    
    while(True):
        while (i <=j and A[j] >= v):
            if A[j] == v:
                A[j], A[e] = A[e], A[j]
                e -= 1
            j -= 1
        while (i <= j and A[i] <= v):
            if A[i] == v:
                A[i], A[s] = A[s], A[i]
                s += 1
            i += 1
         
        if (i > j):
            break
        
        A[i], A[j] = A[j], A[i]
        i += 1
        j -= 1
    
    while(e < n - 1):
        e += 1
        A[e], A[i] = A[i], A[e]
        i += 1
        
    while(s > 0):
        s -= 1
        A[s], A[j] = A[j], A[s]
        j -= 1

    return j + 1, i
    


def kselect(A, k):
    if k < 0 or k >= len(A):
        print("Error: k out of bounds")
        exit(1)
    
    p, q = partitionInPlace(A, A[k])
    
    
    if p <= k and k < q:
        return A[p]
    elif k < p:
        return kselect(A[:p], k)
    elif k >= q:
        return kselect(A[q:], k - q)
        


if __name__ == "__main__":
    
    arr = [random.randint(-1000, 1000) for _ in range(100)]
    
    print(f"Array: {arr}")
    print()
    
    k = 49
    result = kselect(arr, k)
    print(f"The {k}th smallest element is:", result)
    print()
    
    a = sorted(arr)
    print(f"sorted arr[{k}]: {a[k]}")

File: Project_2/Sequential/KSelect.cpp
#include <iostream>

typedef struct Pointers{
    int p;
    int q;
} Pointers;

Pointers partitioninplace(int A[], int v, int left, int right);

void swap(int *a, int *b);

int kselect(int A[], int k, int left, int right);

int main(){

    int arr[] = { 10, 4, 5, 8, 6, 11, 26, 9, 23, 56, 2, 11 }; 
    int n = sizeof(arr) / sizeof(arr[0]); 
    int k = 0; 

    int value = kselect(arr, k, 0, n - 1);

    fprintf(stdout, "The %d-th smallest value is: %d\n", k, value);

    int arr2[] = { 10, 4, 5, 8, 6, 11, 26, 9, 23, 56, 2, 11 }; 
    int m = sizeof(arr2) / sizeof(arr[0]); 

    auto res = partitioninplace(arr2, 0, 0, m-1);
    



    return 0;
}

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}



Pointers partitioninplace(int A[], int v, int left, int right){
    
    Pointers res;

    int i = left;
    int j = right;
    
    int s = i;
    int e = j;
    
    while(1){
        while (i <=j && A[j] >= v){
            if (A[j] == v){
                swap(&A[j], &A[e]);
                e--;
            }
            j--;
        }
        while (i <= j && A[i] <= v){
            if (A[i] == v){
                swap(&A[i], &A[s]);
                s++;
            }
            i++;
        }
        if (i > j)
            break;
        
        swap(&A[i], &A[j]);
        i++;
        j--;
    }
    while(e < right){
        e++;
        swap(&A[i], &A[e]);
        i++;
    }

    while(s > left){
        s--;
        swap(&A[j], &A[s]);
        j--;
    }

    res.p = j + 1;
    res.q = i;

    return res;
 
}

int kselect(int A[], int k, int left, int right){
    int n = right + 1;
    if (k < 0 || k >= n){
        printf("Error: k out of bounds");
        exit(1);
    }
    
    Pointers res = partitioninplace(A, A[k], left, right);

    if (k < res.p){
        return kselect(A, k, left, res.p - 1);
    }  
    else if (k >= res.q){
        return kselect(A, k , res.q, right);
    }
    
    return A[res.p];
}